zkp-prover/src/zkp_component.rs

```diff
@@ -0,0 +1,283 @@
+    zkp_requests: Arc<Mutex<ZKPRequests<N>>>,
+    pub(crate) zkp_events: BroadcastSender<ZKProof>,
+}
+
+impl<N: Network> Clone for ZKPComponentProxy<N> {
+    fn clone(&self) -> Self {
+        Self {
+            network: Arc::clone(&self.network),
+            zkp_state: Arc::clone(&self.zkp_state),
+            zkp_requests: Arc::clone(&self.zkp_requests),
+            zkp_events: self.zkp_events.clone(),
+        }
+    }
+}
+
+impl<N: Network> ZKPComponentProxy<N> {
+    pub fn get_zkp_state(&self) -> ZKPState {
+        self.zkp_state.read().clone()
+    }
+
+    pub fn request_zkp_from_peers(&mut self, peers: Vec<N::PeerId>) -> bool {
+        let mut zkp_requests_l = self.zkp_requests.lock();
+        if zkp_requests_l.is_finished() {
+            zkp_requests_l.request_zkps(peers, self.zkp_state.read().latest_block_number);
+            return true;
+        }
+        false
+    }
+
+    pub fn subscribe_zkps(&self) -> BroadcastStream<ZKProof> {
+        BroadcastStream::new(self.zkp_events.subscribe())
+    }
+}
+
+/// Election blocks ZKP Component. Has:
+///
+/// - The previous election block public keys
+/// - Previous election block header hash
+/// - The pk tree root at genesis or none if genesis block
+///
+/// The proofs are returned by polling the components.
+#[pin_project]
+pub struct ZKPComponent<N: Network> {
+    blockchain: Arc<RwLock<Blockchain>>,
+    network: Arc<N>,
+    zkp_state: Arc<RwLock<ZKPState>>,
+    zk_prover: Option<ZKProver<N>>,
+    zk_proofs_stream: ZKProofsStream<N>,
+    proof_storage: ProofStore,
+    zkp_requests: Arc<Mutex<ZKPRequests<N>>>,
+    zkp_events: BroadcastSender<ZKProof>,
+}
+
+impl<N: Network> ZKPComponent<N> {
+    pub async fn new(
+        blockchain: Arc<RwLock<Blockchain>>,
+        network: Arc<N>,
+        is_prover_active: bool,
+        env: Environment,
+    ) -> Self {
+        // Defaults zkp state to genesis
+        let network_info = NetworkInfo::from_network_id(blockchain.read().network_id());
+        let genesis_block = network_info.genesis_block::<Block>().unwrap_macro();
+
+        let zkp_state = Arc::new(RwLock::new(
+            ZKPState::with_genesis(&genesis_block).expect("Invalid genesis block"),
+        ));
+
+        let (zkp_events, _rx) = broadcast(BROADCAST_MAX_CAPACITY);
+
+        // Load from db the preexisting proof if any
+        let proof_storage = ProofStore::new(env.clone());
+        Self::load_proof_from_db(&blockchain, &zkp_state, &proof_storage, &zkp_events);
+
+        let zk_prover = if is_prover_active {
+            Some(
+                ZKProver::new(
+                    Arc::clone(&blockchain),
+                    Arc::clone(&network),
+                    Arc::clone(&zkp_state),
+                )
+                .await,
+            )
+        } else {
+            None
+        };
+
+        // Sets the stream of the broadcasted
+        let zk_proofs_stream = network.subscribe::<ZKProofTopic>().await.unwrap().boxed();
```

Comment:
Is it ok to `unwrap` here?

Reply:
We unwrap in all other places. So I guess it is ok at least for now